### PR TITLE
fix: cd workflows

### PR DESCRIPTION
**Pull request by francardoso93**
Fixed Broken CD workflows:
- Add missing GH actions permissions to "deploy-job"
- secrets.FIREBASE_CLIENT_NAME doesn't exist. Replaced by secrets.FIREBASE_CLIENT_EMAIL

Extras:
- Renamed "deploy-step" to "deploy-job", as it actually contains a job
- Created a standard name between "cds" for different environments
- Add workflow_dispatch for future feature branches (There is a trade-off: Makes it easier to test feature branches, but also allow people to manually run from "master" branch)
- Fixed Terraform error "Unsupported block type value_from dynamic" by updating provider version

Before approving/merging : 
- Double check if it's OK for Dev environment to use some values that are equal to prod (FIREBASE_CLIENT_EMAIL, FIREBASE_PROJECT_ID, FIREBASE_PRIVATE_KEY_BASE64)
- If that's Ok, uncomment last step of "deploy-job", and remove TODO comment

To fix this issue, work was also required through GCP CLI and console, therefore, here are some suggestions for future improvement: https://github.com/nearform/optic/issues/238#issuecomment-1435156624

Closes https://github.com/nearform/optic/issues/238.

### CODE REVIEW

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
+permissions:
+  id-token: read
+
 on:
   push:
     branches:
       - master
+  pull_request:
+    branches:
+      - master
```
Comment (simoneb) [inline .github/workflows/cd.yml:12]: if you need this for testing, please use a workflow_dispatch event instead, so you don't have to remember to remove the trigger afterwards

**Comment by francardoso93** [comment]
`id-token: write` is required for requesting JWT, as explained at https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#adding-permissions-settings

**Comment by francardoso93** [comment]
Workflow is now failing with a different error:
https://github.com/nearform/optic/actions/runs/4196541905/jobs/7277650867

Invalid value for "audience".

The assumption is that value for secret provided in `workload_identity_provider` is probably wrong.

Please see that in https://github.com/google-github-actions/auth#inputs, it uses the same values a default.

File: .github/workflows/cd-dev.yml [hunk 2]
```diff
@@ -3,16 +3,18 @@ name: cd-dev
 on:
   push:
     branches:
-      - master
-
+      - master  
+  pull_request: ## TODO:Remove after validating
+      branches:
+      - master
```
Comment (simoneb) [inline .github/workflows/cd-dev.yml:9]: use workflow_dispatch please, I had already posted this comment elsewhere

Comment (francardoso93) [reply .github/workflows/cd-dev.yml:9]: workflow_dispatch doesn't work from feature branch, just when configured in main

File: .github/workflows/deploy-job.yml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-name: deploy-step
+name: deploy-job
```
Comment (codeflyer) [inline .github/workflows/deploy-job.yml:1]: Any specific reason to rename this file. 
I agree that `deploy-job` is more suitable as a name, but we use deploy-step in every project.

File: .github/workflows/deploy-job.yml [hunk 4]
```diff
@@ -65,16 +68,15 @@ jobs:
...
           region: ${{ secrets.gcp_cloudrun_service_region }}
           project_id: ${{ secrets.gcp_project_id }}
-          secrets: >-
-            FIREBASE_CLIENT_EMAIL=${{ secrets.firebase_client_email_gcp_secret_name }}:latest,
-            FIREBASE_PROJECT_ID=${{ secrets.firebase_project_id_gcp_secret_name }}:latest,
-            FIREBASE_PRIVATE_KEY_BASE64=${{ secrets.firebase_private_key_base64_gcp_secret_name }}:latest
+          env_vars: >-
+            FIREBASE_CLIENT_EMAIL=${{ secrets.firebase_client_email }},
+            FIREBASE_PROJECT_ID=${{ secrets.firebase_project_id }},
+            FIREBASE_PRIVATE_KEY_BASE64=${{ secrets.firebase_private_key_base64 }}
```
Comment (codeflyer) [inline .github/workflows/deploy-job.yml:80]: Isn't `FIREBASE_PRIVATE_KEY_BASE64` a secret?

Comment (RomuloVitoi) [inline .github/workflows/deploy-job.yml:80]: Looks like the error start appearing after changing the secrets to env vars and then trying to change it back to secrets. Could you try deploying a version without any env vars and then creating it with the secrets?

I think for email and project id it is fine to leave them as env vars, but the private key should be kept a secret.

If we decide to leave the email and project id as env vars then we can remove them from the `update-secrets` step (and cleanup the `_secret_name` variables).

**Review by RomuloVitoi — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy-job.yml (hunk 3 above)
Comment (amanvr) [reply .github/workflows/deploy-job.yml:1]: I'm not sure. This was done by the original author

File: .github/workflows/deploy-job.yml (hunk 4 above)
Comment (amanvr) [reply .github/workflows/deploy-job.yml]: I've updated those now. Let me know if you want to add those secrets back @RomuloVitoi

**Review by RomuloVitoi — APPROVED** [review]

File: .github/workflows/deploy-job.yml (hunk 3 above)
Comment (simoneb) [reply .github/workflows/deploy-job.yml:1]: it's fine, let's keep it as it is

File: .github/workflows/deploy-job.yml [hunk 5]
```diff
@@ -54,27 +53,19 @@ jobs:
...
-            echo "${{ secrets.firebase_client_email }}" | tr -d '\n' | gcloud secrets versions add ${{ secrets.firebase_client_email_gcp_secret_name }} --data-file=-
-          fi
-
-          if [ "${{ secrets.firebase_project_id }}" != "$(gcloud secrets versions access latest --secret ${{ secrets.firebase_project_id_gcp_secret_name }})" ]; then
-            echo "${{ secrets.firebase_project_id }}" | tr -d '\n' | gcloud secrets versions add ${{ secrets.firebase_project_id_gcp_secret_name }} --data-file=-
-          fi
-
           if [ "${{ secrets.firebase_private_key_base64 }}" != "$(gcloud secrets versions access latest --secret ${{ secrets.firebase_private_key_base64_gcp_secret_name }})" ]; then
             echo "${{ secrets.firebase_private_key_base64 }}" | tr -d '\n' | gcloud secrets versions add ${{ secrets.firebase_private_key_base64_gcp_secret_name }} --data-file=-
           fi
```
Comment (simoneb) [inline .github/workflows/deploy-job.yml:58]: I'm ok keeping it this way, but @amanvr can you please create an issue to replace this with https://github.com/nearform-actions/github-action-gcp-secrets in the future?

File: .github/workflows/cd-production.yml
Comment (simoneb) [inline .github/workflows/cd-production.yml:1]: let's keep the original name of this file please

File: .github/workflows/deploy-job.yml (hunk 5 above)
Comment (amanvr) [reply .github/workflows/deploy-job.yml:58]: Created #261

**Review by simoneb — APPROVED** [review]

**Pull request merged**